File: backend/src/jobtrackr_api/models/report.py
from datetime import datetime

from pydantic import BaseModel

from jobtrackr_api.models.tracker import (
    TrackerDeadlineItem,
    TrackerDistributionItem,
    TrackerPipelineGroup,
    TrackerRecentActivityItem,
    TrackerSummaryCard,
)


class CareerReportMetadata(BaseModel):
    report_id: str
    generated_at: datetime
    report_title: str
    period_label: str


class CareerReportExecutiveSummary(BaseModel):
    total_opportunities: int
    active_opportunities: int
    closed_opportunities: int
    applied_count: int
    screening_count: int
    interview_count: int
    offer_count: int
    rejected_count: int


class CareerReportPracticalNotes(BaseModel):
    deterministic_report: bool
    generated_from_manual_opportunities: bool
    no_scraping: bool
    no_external_job_board_ingestion: bool
    no_ai_generated_ranking: bool


class CareerProgressReport(BaseModel):
    metadata: CareerReportMetadata
    executive_summary: CareerReportExecutiveSummary
    summary_cards: list[TrackerSummaryCard]
    status_distribution: list[TrackerDistributionItem]
    source_distribution: list[TrackerDistributionItem]
    priority_distribution: list[TrackerDistributionItem]
    pipeline_groups: list[TrackerPipelineGroup]
    upcoming_deadlines: list[TrackerDeadlineItem]
    overdue_opportunities: list[TrackerDeadlineItem]
    recent_activity: list[TrackerRecentActivityItem]
    practical_notes: CareerReportPracticalNotes

File: frontend/src/types/report.ts
import type {
  TrackerDeadlineItem,
  TrackerDistributionItem,
  TrackerPipelineGroup,
  TrackerRecentActivityItem,
  TrackerSummaryCard,
} from "./tracker";

export type CareerReportMetadata = {
  report_id: string;
  generated_at: string;
  report_title: string;
  period_label: string;
};

export type CareerReportExecutiveSummary = {
  total_opportunities: number;
  active_opportunities: number;
  closed_opportunities: number;
  applied_count: number;
  screening_count: number;
  interview_count: number;
  offer_count: number;
  rejected_count: number;
};

export type CareerReportPracticalNotes = {
  deterministic_report: boolean;
  generated_from_manual_opportunities: boolean;
  no_scraping: boolean;
  no_external_job_board_ingestion: boolean;
  no_ai_generated_ranking: boolean;
};

export type CareerProgressReport = {
  metadata: CareerReportMetadata;
  executive_summary: CareerReportExecutiveSummary;
  summary_cards: TrackerSummaryCard[];
  status_distribution: TrackerDistributionItem[];
  source_distribution: TrackerDistributionItem[];
  priority_distribution: TrackerDistributionItem[];
  pipeline_groups: TrackerPipelineGroup[];
  upcoming_deadlines: TrackerDeadlineItem[];
  overdue_opportunities: TrackerDeadlineItem[];
  recent_activity: TrackerRecentActivityItem[];
  practical_notes: CareerReportPracticalNotes;
};
